### DISCUSSION TITLE
Cutting down boot time in UefiPayloadPkg

### DISCUSSION THREAD

**Discussion by AndrewFalanga** [tianocore/edk2#5466] [Q&A] [answered]
I am working on using the `UefiPayloadPkg` with Slimbootloader. SBL is rather quick: ~4 seconds. How can I trim time away from the UEFI payload in order to keep the hand off to GRUB under 2 or 3 seconds? The `UefiPayloadPkg.dsc` file already seems quite streamlined when I compare it with the `OvmfPkgX64.dsc` file. (Prior to testing on hardware, I was using QEMU).

What software packages/libraries built into UEFI can be tweaked in order to achieve faster boot times? The Googling that I've done thus far has much, much to say about tweaking settings in a BIOS/UEFI implementation. These _all_ relate to an existing build of UEFI. How can one streamline their build of UEFI?

**Comment by marcjones-syspro** [comment, marked as answer]
In my experience most of the time is taken in device driver loading and devie enumeration. The more you limit the device support the faster it will get. You can remove support for some devices, like NVME or USB in the payload .dsc. https://github.com/tianocore/edk2/blob/master/UefiPayloadPkg/UefiPayloadPkg.dsc

You could also look at implementing a fastboot option that doesn't re-scan for new devices in the BDS. 

It would be nice to see some improvements to the core UefiPayload. Dasharo and Mr Chromebox have done most of the more recent work with UefiPaylaod with coreboot, but it hasn't all been pushed back upstream.  

https://github.com/Dasharo/edk2
https://github.com/MrChromebox/edk2

**Reply by AndrewFalanga** [reply]
Thank you! I have recently completed writing a PCIe driver for UEFI and I would tend to agree with this. The driver model is nice, but it does have inefficiency built in. Thanks for the pointers.
